### ISSUE TITLE
pcloud cleanup error, log in required

### ISSUE THREAD

**Issue by jollycar** [rclone/rclone#3853]
## What is the problem you are having with rclone?

I get an error when trying to cleanup my pCloud remote.

Note: I succesfully created a pcloud remote and can sync files without problems.

## What is your rclone version (output from `rclone version`)

rclone v1.50.2
- os/arch: linux/amd64
- go version: go1.13.4

Note: I tried the latest beta (rclone v1.50.2-103-g7e6fac8b-beta), but the problem is still there

## Which OS you are using and how many bits (eg Windows 7, 64 bit)

Linux Ubuntu 18.04 64 bit

##  Which cloud storage system are you using? (eg Google Drive)

pCloud

## The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

"rclone cleanup pcloud:"

## A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

"rclone cleanup -vv pcloud:
2020/01/04 12:46:57 DEBUG : rclone: Version "v1.50.2-103-g7e6fac8b-beta" starting with parameters ["rclone" "cleanup" "-vv" "pcloud:"]
2020/01/04 12:46:57 DEBUG : Using config file from "/home/ivo/.config/rclone/rclone.conf"
2020/01/04 12:46:57 ERROR : Attempt 1/3 failed with 1 errors and: pcloud error: Log in required. (1000)
2020/01/04 12:46:58 ERROR : Attempt 2/3 failed with 1 errors and: pcloud error: Log in required. (1000)
2020/01/04 12:46:58 ERROR : Attempt 3/3 failed with 1 errors and: pcloud error: Log in required. (1000)
2020/01/04 12:46:58 Failed to cleanup: pcloud error: Log in required. (1000)
"

**Comment by ncw** [comment]
Here is the API method: https://docs.pcloud.com/methods/trash/trash_clear.html

And here is how rclone is calling it

```
2020/01/04 13:30:43 DEBUG : HTTP REQUEST (req 0xc000551800)
2020/01/04 13:30:43 DEBUG : POST /trash_clear?folderid=0 HTTP/1.1
Host: api.pcloud.com
User-Agent: rclone/v1.50.2-104-gc018ebeb-fix-3851-appdata-listr-beta
Content-Length: 0
Authorization: XXXX
Accept-Encoding: gzip
```

```
2020/01/04 13:30:44 DEBUG : HTTP RESPONSE (req 0xc000551800)
2020/01/04 13:30:44 DEBUG : HTTP/1.1 200 OK
Connection: close
Content-Length: 49
Cache-Control: private, max-age=0
Content-Type: application/json; charset=utf-8
Date: Sat, 04 Jan 2020 13:30:43 GMT
Etag: "cp9FcrQXC7LHQdeUPdbwDByU9hf7"
Server: CloudHTTPd-API v1.1
Vary: Accept-Encoding
X-Error: 1000

{
	"result": 1000,
	"error": "Log in required."
}
```

I don't really understand why that doesn't work - rclone is providing the `Authorization` header...

Do you have support with pcloud?  Maybe you could ask them?

**Comment by ivohuisjes** [comment]
Where do you get the POST url path "/trash_clear" from? According to the API docs it should be "/getfilepublink"?

**Comment by ivohuisjes** [comment]
I also send this support request to pcloud: 

> I am using a third party application (rclone) to communicate with pcloud, using the REST API. There is a problem using the API's "trash_clear" (https://docs.pcloud.com/methods/trash/trash_clear.html) command. When rclone does a POST request to /trash_clear, the error response is 1000 (Log in required.). I noticed the URL on the API page is /getfilepublink, but rclone uses /trash_clear. Is this a problem in your documentation or a problem by clone?

**Comment by ivohuisjes** [comment]
Hello,
Thank you for the e-mail!

Unfortunately, pCloud does not support rclone. The problem might occur because of it.

Kind regards,
Jonathan Teuling,
pCloud Support Team

**Comment by ncw** [comment]
> Where do you get the POST url path "/trash_clear" from? According to the API docs it should be "/getfilepublink"?

I guess that was in the docs in a previous version...

It looks like a more sensible URL than `getfilepublink` which is described here: https://docs.pcloud.com/methods/public_links/getfilepublink.html

> Unfortunately, pCloud does not support rclone. The problem might occur because of it.

That wasn't very helpful!

Maybe you could emphasize that you are trying to use their public API?

**Comment by accountForIssues** [comment]
@ncw I did some tests (using curl and python requests) and it looks like the `Bearer` token provided by the OAuth process (which rclone uses) is not accepted by certain methods (all the trash ones, some auth ones, etc.).

I couldn't find any mention of this in their documentation so I assume either it's a bug that likely was introduced when they updated their api.

If, however, we use the `getauth` method with 'username/password' combo, this token is accepted and `trash_list`, `trash_clear`, etc. complete successfully.

**Comment by accountForIssues** [comment]
Update: I received a reply from pCloud.

> Hello,
> 
> Thank you for taking the time to contact pCloud's Technical Support. My name is Stefan and I will be happy to assist you.
> 
> Please excuse us for the inconvenience but note that we do not provide dedicated API support so I'm unable to advise you on how to use the code.
> The information we can provide is only at the link below as you have already found out yourself.
> https://docs.pcloud.com/

That sucks and brings down my opinion of pCloud.

So, the option is for rclone to switch to `getauth` instead of using `OAuth` for all the methods to work.

**Comment by ncw** [comment]
> I did some tests (using curl and python requests) and it looks like the `Bearer` token provided by the OAuth process (which rclone uses) is not accepted by certain methods (all the trash ones, some auth ones, etc.).
> 
> I couldn't find any mention of this in their documentation so I assume either it's a bug that likely was introduced when they updated their api.

Sounds likely

> If, however, we use the `getauth` method with 'username/password' combo, this token is accepted and `trash_list`, `trash_clear`, etc. complete successfully.

Interesting.

`getauth` has the disadvantage that rclone needs to store the password in the config file. It would hardly be the first backend to do that, but it is annoying since oauth already works nearly!

What did you say in your email? Maybe you should rephrase it as a bug report that the clear trash API works with a token from `getauth` but not from OAuth - that might get some traction? Or maybe you did this already!

**Comment by accountForIssues** [comment]
> `getauth` has the disadvantage that rclone needs to store the password in the config file. It would hardly be the first backend to do that, but it is annoying since oauth already works nearly!

I forget what the expiry was for the `getauth` token but storing the password could be optional. Let the user know that storing the password would auto refresh the token when needed failing which the user needs to keep an eye in the logs.

But, yeah, I wouldn't mind storing the user/pass if required. As you said, not the only remote that does this.

> What did you say in your email? Maybe you should rephrase it as a bug report that the clear trash API works with a token from `getauth` but not from OAuth - that might get some traction? Or maybe you did this already!

Yeah, I said exactly that. I explained the bug and even the errors in the documentation. Maybe they use another API or the `getauth` method for their in-house clients hence they haven't faced issues themselves, I dunno.

I'm surprised they didn't even acknowledge the clear mistake about `trash_clear` method where it says `getfilepublink` in the documentation.

**Comment by ncw** [comment]
@accountForIssues do you want to have a go at adding the password into the config and using it for those methods?

Remember the code will need to be backwards compatible with no password set.

**Comment by accountForIssues** [comment]
@ncw I have never programmed using Go but I'm definitely interested in learning it. This could be a good way to learn by example. I would like to take a stab at it but cannot promise anything.

**Comment by ncw** [comment]
@accountForIssues let me know if you need any help. I can give lots of pointers if required :-)

**Comment by ambipur76** [comment]
Are there any news related to this problem? Using pcloud for regular backups it is crucial to be able do empty the rash from the command line.

**Comment by ncw** [comment]
@ambipur76 we are still looking for a volunteer to do the patch. Might you be interested?

**Comment by ambipur76** [comment]
I successfully installed go and downloaded rclone-master.zip. After extracting it I managed to compile rclone with the go build command.
In /backend/pcloud/pcloud.go I have found the function
func (f *Fs) CleanUp(ctx context.Context)
I suppose this should be modified. In previous posts you have mentioned that getauth should be used and a disadvantage because password should be than stored in rclone.conf.
At this point I would need some more concrete tips or support to be able to proceed further.
